### ISSUE TITLE
Doobie + Hikari + Redshift performance 

### ISSUE THREAD

**Issue by mirosval** [tpolecat/doobie#657]
I am trying to use this combination in a microservice that is connecting to Redshift and executing only select queries. One request to the microservice results in multiple SQL queries to the database. I was unable to locate an example to work off of, and I'm probably doing something wrong, but I can't figure out what it is.

I initialize my hikari Transactor like so:
```scala
val transactorFuture: Task[HikariTransactor[Task]] = for {
      _ <- Async[Task].delay(Class.forName(driver()))
      t <- HikariTransactor.initial[Task]
      _ <- t.configure { ds =>
        ds.setJdbcUrl(databaseUrl())
        ds.setUsername(databaseUser())
        ds.setPassword(databasePass())
        ds.setMaximumPoolSize(databaseConnectionPoolSize())
        ds.setPoolName("hikari-pool")
        ds.setAutoCommit(false)
        Task.delay(ds)
      }
    } yield t
```

and I perform requests like so:
```scala
val sql              = request.sql
val query: Query0[T] = sql.query[T]
val start            = System.nanoTime()
val strategy = Strategy(unit, unit, unit, close)

(strategy.wrap andThen transactor.rawTrans)
      .apply(query.unique)
      .map { result =>
        logger.info(s"Fetch Unique for $sql took ${(System.nanoTime() - start) / 1e6}ms")
        result
      }
      .runAsync(scheduler)
```

and this is what I see in the logs:
```
17:17:27.126 [db-io-pool-54] INFO  doobie.util.log$LogHandler$ - Successful Statement Execution:

  select sum( a ) from b where c

 arguments = []
   elapsed = 105 ms exec + 0 ms processing (106 ms total)

17:17:27.190 [db-io-pool-33] INFO  n.c.s.r.database.SQLDatabase - Fetch Unique for Fragment("select sum( a ) from b where c") took 246.1052ms
```

where did ~140ms go? (246.1052ms - 106 ms)

Redshift recommends not committing, so I modified the default Transactor Strategy to avoid commit, but if I also eliminate the connection close after the query I start leaking connections.

I am assuming that the 100ms went to establishing a connection? How can I reuse connections then?

It is very difficult to figure out how much time is spent where inside of the Doobie free monad and I have spent almost a week hunting for answers before opening this issue.

**Comment by tpolecat** [comment]
First be sure that you're only constructing a single `HikariTransactor` and passing it to the rest of your program, since each one you construct will have a new pool in its internal state (i.e., `transactorFuture` should not appear in the code that runs per-request.

You can replace the `Strategy` in your transactor via

```scala
val newXA = Transactor.strategy.set(oldXA, newStrategy)
```

and then `query.unique.transact(xa)` so you're not dealing with the natural transformations directly.

doobie is reporting the statement execution time. You're adding in the time it takes to construct the program, schedule execution, and acquire/release the connection. If you're constructing a new transactor this could certainly account for the extra time.

**Comment by mirosval** [comment]
Thank you for your super fast and comprehensive response! 

It may not be apparent from the snippets, but I'm pretty sure I only construct one Hikari Transactor. 

What I don't fully understand is why it is not possible to reuse connections for multiple SQL queries. Doobie always starts with `getConnection()`, but that seems to always create a new connection instead of reusing an already open connection. 

If I have a connection pool of 10 connections as per the Hikari best practices, after 10 SQL queries the subsequent ones always time out (presumably because they couldn't get a new connection). Only if I use a strategy that closes connections everything works as expected.

I suspect that the 100ms goes to opening the connection, setting autocommit etc.

**Comment by tpolecat** [comment]
Connections are re-used (that's the purpose of a connection pool) … "closing" a connection just resets some things and returns it to the pool, which should be a very fast operation. You mentioned that you turned commit off, but you also have auto-commit off which means that closing the connection without commit may initiate a rollback. So you might try turning auto-commit **on**. I would also turn on driver logging (see the Redshift JDBC documentation) to get more information about what's actually going on at the wire level.

**Comment by mirosval** [comment]
I'll do that and come back with the results, thank you for your amazing support! 🎆

**Comment by mirosval** [comment]
@tpolecat I have turned up logging from the Redshift driver and posted the logs here:
https://gist.github.com/mirosval/4d34f703bfed3b057d813eb01db44854

Both files are from the same JVM so you can directly compare the times logged.

It seems that the connections are getting reused, but the Redshift driver exchanges data with the DB 4x at ~60ms each (`prepareStatement`, `sendParams`, `executeStatement`, `close`). This is running from my local machine against AWS, I have also tried deploying it on AWS, in which case it was about 20ms per round trip.

It seems to me that we should be able to speed it up by reusing prepared statements?

**Comment by tpolecat** [comment]
When you call `.transact` you get back a program that runs on a logically new connection. doobie's high level API ensures that things are cleaned up eagerly, and the connection pool ensures that everything is reset so you can't tell it's really a long-lived connection.

If you want to re-use statements (which are valid only for the lifetime of a connection) you may need to create a statement pool that lets you re-use the same set of statements across multiple requests to your application. This would cut your message exchanges in half.

I'm not sure how I would do this off the top of my head, but an IO effect feeding an fs2 queue might be a place to start. You need a way to clean up connections that get invalidated, time out, etc., so there may be some tricky error handling to deal with.

I'll be interested to see what you come up with.

**Comment by mirosval** [comment]
@tpolecat So today I woke up thinking I am going to write my own caching DataSource and build up a Transactor out of that. Then I read this:
https://github.com/brettwooldridge/HikariCP#statement-cache
They specifically say it is a bad idea, but they also say that most drivers just support caching prepared statements out of the box. 

So I swapped the Redshift JDBC driver for Postgres JDBC driver and now everything is **4x faster** (after some initially slow queries while the prepared statements are cached). It was a 2-line change and no changes to Doobie are required.

I figured I'd leave this info here, because I suspect others will run into this issue as well.

Thank you again for your ideas and tips. They really helped. :)
